### PR TITLE
Backbones refactor

### PR DESCRIPTION
**Pull request by klemen1999**
- Renamed some backbone files
- Made BaseBackbone class
- EfficientRep now always returns all 4 feature maps and doesn't take `is_4heads` as parameter anymore
- Added backbone inference tests

### CODE REVIEW

File: src/luxonis_train/models/backbones/micronet.py [hunk 1]
```diff
@@ -21,82 +228,68 @@ def __init__(self):
...
-        c2,
-        init_a=[0.0, 0.0],
-        init_b=[0.0, 0.0],
-        act_relu=True,
-        g=None,
-        reduction=4,
-        expansion=False,
+        in_channels: int,
+        out_channels: int,
+        init_a: Optional[list] = [0.0, 0.0],
```
Comment (kozlov721) [inline src/luxonis_train/models/backbones/micronet.py:255]: Better to get rid of mutable default arguments. Can it be a tuple instead? Or if not, then
```python
init_a: Optional[list] = None
# later in __init__
init_a = init_a or [0.0, 0.0]
```

File: src/luxonis_train/models/necks/reppan_neck.py [hunk 2]
```diff
@@ -41,7 +41,8 @@ def __init__(
         ]
 
         self.is_4head = is_4head
-        prev_out_start_idx = 1 if self.is_4head else 0
+        # prev_out_start_idx = 1 if self.is_4head else 0
```
Comment (kozlov721) [inline src/luxonis_train/models/necks/reppan_neck.py:44]: Can we delete the commented code?

File: src/luxonis_train/models/backbones/mobileone.py [hunk 3]
```diff
@@ -6,12 +6,167 @@
...
+        assert (
+            variant in MOBILEONE_VARIANTS_SETTINGS.keys()
+        ), f"MobileOne model variant should be in {list(MOBILEONE_VARIANTS_SETTINGS.keys())}"
+
+        variant_params = MOBILEONE_VARIANTS_SETTINGS[variant]
+        self.backbone = MobileOne_(**variant_params)
+        self.is_inference = False
+
+    def forward(self, x):
+        # TODO: when to perform reparameterization?
```
Comment (kozlov721) [inline src/luxonis_train/models/backbones/mobileone.py:37]: I think it would make sense to have `to_deploy` with the reparametrization.

File: src/luxonis_train/models/necks/reppan_neck.py (hunk 2 above)
Comment (klemen1999) [reply src/luxonis_train/models/necks/reppan_neck.py:44]: This will be addresse in separate commit where whole neck is refactored.

File: src/luxonis_train/models/necks/reppan_neck.py [hunk 4]
```diff
@@ -147,10 +148,11 @@ def __init__(
             )
 
     def forward(self, x):
-        if self.is_4head:
-            x3, x2, x1, x0 = x
-        else:
-            x2, x1, x0 = x
+        x3, x2, x1, x0 = x
+        # if self.is_4head:
```
[the remark this answers is not in the document]
Comment (klemen1999) [reply src/luxonis_train/models/necks/reppan_neck.py:152]: This will be addresse in separate commit where whole neck is refactored.

File: src/luxonis_train/models/backbones/mobileone.py (hunk 3 above)
Comment (klemen1999) [reply src/luxonis_train/models/backbones/mobileone.py:37]: Yeah you are right, will change this.

File: src/luxonis_train/models/backbones/micronet.py (hunk 1 above)
Comment (klemen1999) [reply src/luxonis_train/models/backbones/micronet.py:255]: I see your point but maybe we enforce this only on top level? So on classes which user can actually access (in this case it would be MicroNet class). If we enforce it everywhere it would introduce a lot of extra type conversions or extra code which doesn't really change anything from user's or developer's perspective. Thoughts?

Comment (kozlov721) [reply src/luxonis_train/models/backbones/micronet.py:255]: Would make sense to enforce this it at least for modules that are expected to be instantiated multiple times (I guess `DYShiftMax` would be one of those). Not really necessary for backbones and necks I'd say.

File: src/luxonis_train/models/backbones/contextspatial.py [hunk 5]
```diff
@@ -86,15 +113,15 @@ def forward(self, x):
 
 
 class FeatureFusionModule(nn.Module):
```
Comment (tersekmatija) [inline src/luxonis_train/models/backbones/contextspatial.py:115]: Seems perhaps general enough to be put into Modules?

File: src/luxonis_train/models/backbones/contextspatial.py [hunk 6]
```diff
@@ -68,14 +95,14 @@ def forward(self, x):
 
 
 class AttentionRefinmentModule(nn.Module):
```
Comment (tersekmatija) [inline src/luxonis_train/models/backbones/contextspatial.py:97]: Seems perhaps general enough to be put into Modules?

File: src/luxonis_train/models/backbones/micronet.py [hunk 7]
```diff
@@ -120,178 +313,123 @@ def forward(self, x: Tensor) -> Tensor:
...
+        if min_value is None:
+            min_value = divisor
+        new_v = max(min_value, int(v + divisor / 2) // divisor * divisor)
+        # Make sure that round down does not go down by more than 10%.
+        if new_v < 0.9 * v:
+            new_v += divisor
+        return new_v
+
 
 class SwishLinear(nn.Module):
```
Comment (tersekmatija) [inline src/luxonis_train/models/backbones/micronet.py:333]: Since it's an activation function, would it make sense to move it to modules.activations.py or similar?

**Review by tersekmatija — APPROVED** [review]
Looks good, might be worth re-thinking a bit about what is worth putting into modules and what we keep inside the backbone class?

File: src/luxonis_train/models/backbones/micronet.py (hunk 7 above)
Comment (tersekmatija) [reply src/luxonis_train/models/backbones/micronet.py:358]: I think just for consistency might be best to move them now. As we will forget to do so at a later stage. Though I agree for what is currently planned it's unlikely we'll use them elsewhere in the near future.

Comment (klemen1999) [reply src/luxonis_train/models/backbones/micronet.py:358]: Sounds good. I just found another reference to these  two activations [here](https://github.com/osmr/imgclsmob/blob/68335927ba27f2356093b985bada0bc3989836b1/pytorch/pytorchcv/models/common.py#L26-L51) where they refer to them as HSigmoid (hard-sigmoid) and HSwish also. I'll just stick to this names then and move them in separate file.
